src/dhtproto/node/neo/request/Mirror.d

```diff
@@ -65,18 +69,50 @@ public abstract scope class MirrorProtocol_v0
     {
         import ocean.core.array.Mutation : removeShift;
 
-        /// Array of elements in the queue. Must be set before using the queue.
-        private E[]* queue;
+        /// Backing array in which elements are stored. Must be set before using
+        /// the queue.
+        private E[] queue;
 
         /// Maximum allowed size of the queue in bytes.
         private const max_size = 256 * 1024;
 
-        /// Maximum allowed length of the queue.
-        private const max_length = max_size / E.sizeof;
+        /// Maximum allowed elements of type E of the queue.
+        private size_t max_elems;
+
+        /// Current number of elements in the queue.
+        private size_t num_elems;
+
+        /// Index (E-based) of the next element to write.
+        private size_t write_to_elem;
+
+        /// Index (E-based) of the next element to read.
+        private size_t read_from_elem;
 
+        /// Tests for basic sanity of num_elems and the read/write indices.
         invariant ( )
         {
-            assert(this.queue !is null);
+            if ( this.num_elems == 0 )
+                assert(this.write_to_elem == this.read_from_elem);
+            else
+                assert(this.write_to_elem ==
+                    (this.read_from_elem + this.num_elems) % this.max_elems);
+        }
+
+        /***********************************************************************
+
+            Initialises the queue for use, setting up the provided array for use
+            as the queue's backing storage.
+
+            Params:
+                buf = array to use to back queue storage
+
+        ***********************************************************************/
+
+        public void initialise ( void[]* buf )
+        {
+            this.max_elems = (max_size / E.sizeof) * E.sizeof;
+            buf.length = this.max_elems * E.sizeof;
+            this.queue = (cast(E*)buf.ptr)[0 .. this.max_elems];
```

Comment:
Suggest
- `ref void[] buf` now that you don't store a pointer to the array any more,
- `(cast(E[])*buf)[0 .. this.max_elems]` (without the `*` for `ref void[] buf`).

Reply:
Actually `this.queue = cast(E[])buf;` should be sufficient

Reply:
Updated.